### PR TITLE
[memprof] Add computeUndriftMap

### PR DESCRIPTION
**Pull request by kazutakahirata**
This patch adds computeUndriftMap, a function to compute mappings from
source locations in the MemProf profile to source locations in the IR.

### CODE REVIEW

File: llvm/lib/Transforms/Instrumentation/MemProfiler.cpp [hunk 1]
```diff
@@ -856,6 +857,35 @@ memprof::extractCallsFromIR(Module &M, const TargetLibraryInfo &TLI) {
   return Calls;
 }
 
+DenseMap<uint64_t, LocToLocMap>
+memprof::computeUndriftMap(Module &M, IndexedInstrProfReader *MemProfReader,
+                           const TargetLibraryInfo &TLI) {
+  DenseMap<uint64_t, LocToLocMap> UndriftMaps;
+
+  auto CallsFromProfile = MemProfReader->getMemProfCallerCalleePairs();
+  auto CallsFromIR = extractCallsFromIR(M, TLI);
```
Comment (snehasish) [inline llvm/lib/Transforms/Instrumentation/MemProfiler.cpp:866]: I think spelling out the types here would help the reader.

File: llvm/include/llvm/Transforms/Instrumentation/MemProfiler.h [hunk 2]
```diff
@@ -66,6 +67,21 @@ namespace memprof {
...
 extractCallsFromIR(Module &M, const TargetLibraryInfo &TLI);
 
+struct LineLocationHash {
+  uint64_t operator()(const LineLocation &Loc) const {
+    return Loc.getHashCode();
+  }
+};
+
+using LocToLocMap =
+    std::unordered_map<LineLocation, LineLocation, LineLocationHash>;
```
[the remark this answers is not in the document]
Comment (kazutakahirata) [reply llvm/include/llvm/Transforms/Instrumentation/MemProfiler.h:77]: My laziness.  I would need to define `DenseMapInfo<LineLocation>`, setting aside the empty and tombstones keys.  I'd like to leave this to a follow-up patch.

File: llvm/unittests/Transforms/Instrumentation/MemProfUseTest.cpp [hunk 3]
```diff
@@ -298,4 +304,189 @@ attributes #2 = { builtin allocsize(0) }
...
+  MIB.TotalLifetimeAccessDensity = 23;
+  return MIB;
+}
+
+IndexedMemProfRecord
+makeRecordV2(std::initializer_list<::llvm::memprof::CallStackId> AllocFrames,
+             std::initializer_list<::llvm::memprof::CallStackId> CallSiteFrames,
+             const MemInfoBlock &Block, const memprof::MemProfSchema &Schema) {
+  llvm::memprof::IndexedMemProfRecord MR;
+  for (const auto &CSId : AllocFrames)
```
[the remark this answers is not in the document]
Comment (kazutakahirata) [reply llvm/unittests/Transforms/Instrumentation/MemProfUseTest.cpp]: Fixed in the latest iteration.

File: llvm/unittests/Transforms/Instrumentation/MemProfUseTest.cpp [hunk 4]
```diff
@@ -298,4 +304,189 @@ attributes #2 = { builtin allocsize(0) }
...
+    // We don't populate IndexedAllocationInfo::CallStack because we use it only
+    // in Version1.
+    MR.AllocSites.emplace_back(::llvm::SmallVector<memprof::FrameId>(), CSId,
```
[the remark this answers is not in the document]
Comment (kazutakahirata) [reply llvm/unittests/Transforms/Instrumentation/MemProfUseTest.cpp]: Fixed in the latest iteration.

File: llvm/unittests/Transforms/Instrumentation/MemProfUseTest.cpp [hunk 5]
```diff
@@ -298,4 +304,189 @@ attributes #2 = { builtin allocsize(0) }
...
+                               Block, Schema);
+  for (const auto &CSId : CallSiteFrames)
+    MR.CallSiteIds.push_back(CSId);
+  return MR;
+}
+
+static const auto Err = [](Error E) {
+  consumeError(std::move(E));
+  FAIL();
```
[the remark this answers is not in the document]
Comment (kazutakahirata) [reply llvm/unittests/Transforms/Instrumentation/MemProfUseTest.cpp]: Fixed in the latest iteration.

**Review by kazutakahirata — COMMENTED** [review]
Updated the patch.  Please take a look.  Thanks!

**Review by snehasish — APPROVED** [review]
lgtm

**Pull request merged**